cpp/src/parquet/arrow/reader.cc

```diff
@@ -1233,6 +1274,123 @@ Status FileReaderImpl::ReadRowGroups(const std::vector<int>& row_groups,
   return Status::OK();
 }
 
+struct AsyncBatchGeneratorState {
+  ::arrow::internal::Executor* io_executor;
+  ::arrow::internal::Executor* cpu_executor;
+  std::vector<std::shared_ptr<ColumnReaderImpl>> column_readers;
+  std::queue<std::shared_ptr<RecordBatch>> overflow;
+  std::shared_ptr<::arrow::Schema> schema;
+  int64_t batch_size;
+  int64_t rows_remaining;
+  bool use_threads;
+  bool allow_sliced_batches;
+};
+
+class AsyncBatchGeneratorImpl {
+ public:
+  explicit AsyncBatchGeneratorImpl(std::shared_ptr<AsyncBatchGeneratorState> state)
+      : state_(std::move(state)) {}
+  Future<std::shared_ptr<RecordBatch>> operator()() {
+    if (!state_->overflow.empty()) {
+      std::shared_ptr<RecordBatch> next = std::move(state_->overflow.front());
+      state_->overflow.pop();
+      return next;
+    }
+
+    if (state_->rows_remaining == 0) {
```

Comment:
So the empty rowgroup is handled in two ways:
1. If it's in the middle of a File, the rowgroup could handled by `ColumnReaderImpl`, which skip that rowgroup
2.If it's in the end of a File, it would return a `Finish`?

Am I right?

Reply:
Yes.

> If it's in the end of a File, it would return a Finish?

This can also happen in the middle of a file if the user does not ask for all row groups.

The scan node always calls ReadRowGroupsAsync with a single row group so it is always a possibility (but this is ok because the scan node can handle it).